### ISSUE TITLE
Connections waiting

### ISSUE THREAD

**Issue by pserrao** [brettwooldridge/HikariCP#1233]
#### Environment
```
HikariCP version: x3.2.0
JDK version     : 1.8.0_131
Database        : MySQL
Driver version  : mysql-connector-java-5.1.43-bin
```

-----------------------------------------------------------------------------------------
*Have you searched the CLOSED issues already? How about checking [stackoverflow](https://stackoverflow.com/search?q=HikariCP)?*

Issue with connections waiting - we have a complex environment of an apache server on linux frontending 4 tomcat instances on windows (load balanced with mod_jk). The database is MySQL from Percona (running on linux Centos 7 with 64 cores and 128 GB of memory). Database performance ie great - 99% of transactions occur in milliseconds). The application is a single page application built on spring framework with mybatis controlling the db access.
We are running load tests on Load Impact and monitoring performance with New Relic.
We are finding slowness in the web transactions and a review of New Relic traces points to 80-90% of each transaction spent waiting for a hikari connection.

Tomcat config is as follows: (password and url obscured)
   Resource name="jdbc/EzRegistration" 
        auth="Container" 
		type="javax.sql.DataSource"
    	factory="com.zaxxer.hikari.HikariJNDIFactory"
    	dataSource.user="root"
    	dataSource.password="xxxxxxxxx!"
		dataSource.cachePrepStmts="true"
		dataSource.prepStmtCacheSize="250"
		dataSource.prepStmtCacheSqlLimit="2048"
		dataSource.useServerPrepStmts="true"
		dataSource.useLocalSessionState="true"
		dataSource.useLocalTransactionState="true"
		dataSource.rewriteBatchedStatements="true"
		dataSource.cacheResultSetMetadata="true"
		dataSource.cacheServerConfiguration="true"
		dataSource.elideSetAutoCommits="true"
		dataSource.maintainTimeStats="false"
		dataSource.interactiveClient="true"
		dataSource.characterEncoding="UTF-8"
		dataSource.useSSL="false"
    	driverClassName="com.mysql.jdbc.Driver"
        maximumPoolSize="16"
		minimumIdle="16"
		leakDetectionThreshold="30000"
    	jdbcUrl="jdbc:mysql://127.0.0.1/test"
	

the hikari logs also show connections waiting:
2018-09-10 20:21:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=7, idle=9, waiting=0)
2018-09-10 20:22:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=5, idle=11, waiting=0)
2018-09-10 20:22:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=5, idle=11, waiting=0)
2018-09-10 20:23:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=9, idle=7, waiting=0)
2018-09-10 20:23:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=15, idle=1, waiting=0)
2018-09-10 20:24:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=9)
2018-09-10 20:24:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=2)
2018-09-10 20:25:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=9)
2018-09-10 20:25:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=11)
2018-09-10 20:26:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=27)
2018-09-10 20:26:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=70)
2018-09-10 20:27:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=85)
2018-09-10 20:27:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=131)
2018-09-10 20:28:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=109)
2018-09-10 20:28:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=121)
2018-09-10 20:29:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=188)
2018-09-10 20:29:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=145)
2018-09-10 20:30:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=135)
2018-09-10 20:30:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:31:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=223)
2018-09-10 20:31:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=181)
2018-09-10 20:32:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:32:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=233)
2018-09-10 20:33:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=227)
2018-09-10 20:33:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:34:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=231)
2018-09-10 20:34:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:35:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=15, idle=1, waiting=234)
2018-09-10 20:35:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:36:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:36:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=232)
2018-09-10 20:37:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=233)
2018-09-10 20:37:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=233)
2018-09-10 20:38:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:38:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:39:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=234)
2018-09-10 20:39:44 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=233)
2018-09-10 20:40:14 DEBUG HikariPool:411 - HikariPool-3 - Pool stats (total=16, active=16, idle=0, waiting=231)

the 4 tomcat engines run about 80 connections into the database - from prior experience we find that keeping the connections close to the number of cores is the best methodology.

Can anyone comment on the configuration and see if we are doing something wrong?
Also would moving to mysql-connector-java-8.0.12 help? Does the datasource configuration now work with mysql using this driver version (you can see we use the jdbcurl method as recommended for mysql).

Brent- do you do consulting on Hikari Configuration?

**Comment by pserrao** [comment]
For some reason teh Tomcat configuration did not flow:
Tomcat config is as follows: (password and url obscured)
   <Resource name="jdbc/EzRegistration" 
        auth="Container" 
		type="javax.sql.DataSource"
    	factory="com.zaxxer.hikari.HikariJNDIFactory"
    	dataSource.user="root"
    	dataSource.password="xxxxxxxxx!"
		dataSource.cachePrepStmts="true"
		dataSource.prepStmtCacheSize="250"
		dataSource.prepStmtCacheSqlLimit="2048"
		dataSource.useServerPrepStmts="true"
		dataSource.useLocalSessionState="true"
		dataSource.useLocalTransactionState="true"
		dataSource.rewriteBatchedStatements="true"
		dataSource.cacheResultSetMetadata="true"
		dataSource.cacheServerConfiguration="true"
		dataSource.elideSetAutoCommits="true"
		dataSource.maintainTimeStats="false"
		dataSource.interactiveClient="true"
		dataSource.characterEncoding="UTF-8"
		dataSource.useSSL="false"
    	driverClassName="com.mysql.jdbc.Driver"
        maximumPoolSize="16"
		minimumIdle="16"
		leakDetectionThreshold="30000"
    	jdbcUrl="jdbc:mysql://127.0.0.1/test"
	/>

**Comment by brettwooldridge** [comment]
@pserrao Unfortunately, I do not have sufficient time to do consulting re: HikariCP.

Having said that, I'll do what I can to help.  It seems to me that 200+ threads on the frontend is likely excessive.  You can swap out the pool for another pool as a double-check on this supposition.  Without much to back it up, except intuition, my guess is that frontend threads should be roughly 2.5x the pool size, e.g. 40 threads.

I do suggest enabling metrics.  Either dropwizard or micrometer+???.  If you can collect metrics over the run, all of them polled at a 1sec interval, that would be useful to see.

**Comment by amanullah92** [comment]
@pserrao - Can you share some more information regarding how much TPS you are hitting, how many QPS you are running on the database. Are you keeping the transaction open for a long time while doing some other computation, leading to connection being kept in use for far longer than it should?

Since I think you are doing it in performance testing, I assume you can reliably reproduce the situation. I would suggest taking multiple thread dumps just before the pool begins to timeout- when active connections begin to increase. I've seen in the past that threads can block sometimes due to logging/GC issues, so better check these.

**Comment by billoneil** [comment]
- What do the database metrics look like is it under load?
- Check the p99's of the endpoints and SQL queries or if you open newrelic you take a look at the max query times. A few slower queries could saturate the pools even if 99% of them respond in milliseconds.
- If you have ~80 connections fully utilized to the db you should be getting some really high throughput if they are all only a few ms. Is it lower than expected?

If most of the queries are in the low ms range lets say 10ms and you have 80 connections you should be getting ~8,000 transactions per second. If it's nowhere close to that you probably have some slow queries that are being averaged away by the fast queries.

I'd agree with @brettwooldridge it seems like you have too many connections being opened to the servers and they are not responding fast enough. It just so happens that HikariCP is the most restrictive with resources which is why it's the bottleneck. If you reduce the number of connections the webserver can process it will push some of the queuing up the chain to the webserver and reduce the contention on HikariCP but it probably wouldn't improve performance without fixing the underlying cause.

**Comment by rabindrarakshit** [comment]
@pserrao It would be interesting to know how did you fix the above issue of growing wait connections, we are facing the same issue now. 
@brettwooldridge  any updates you want to provide after 2 years now for the above issue?

**Comment by brettwooldridge** [comment]
@RabindraRakshit I think @billoneil point above is extremely relevant.  Even if 99% of queries run in single-digit milliseconds, if there is one or two that are commonly executed that take, for example, 100ms then that becomes the bottleneck.

It is extremely important to have the number of front-end connections properly matched to the backend SQL server CPU cores.  The common formula is that connections should be 2.5x the number of CPU cores on the database server, *not counting hyper-threading.*

**Comment by aifex** [comment]
@rabindrarakshit , if this might help, I was facing this issue recently and it gave me a tough time.

As has been explained in the previous comments, the server accepts too many requests which hikari pool is not able to handle and so hikari tries to obtain extra connections to serve this spike in demand. 

Eg. For `tomcat` with 200 default threads, if your maxPoolSize = 10, on spike demands, your server would try to serve 200 threads at the same time. If the connections in the pool are busy, hikari would try to obtain 190 connections and this is what you see in the `waiting`.

Here is how I am managing it.
- I made sure that tomcat threads do not exceed the number of hikari maxPoolSize. In that way, there won't be need to ask for more connections during spike.

in spring boot, this is the config I used.  
`server.tomcat.threads.max = 50`
`spring.datasource.hikari.maximumPoolSize = 50`

Note: 50 is variable based on your server capacity
